slang-checks/src/main/java/org/sonarsource/slang/checks/UnusedPrivateMethodCheck.java

```diff
@@ -52,27 +48,27 @@
   @Override
   public void initialize(InitContext init) {
     init.register(ClassDeclarationTree.class, (ctx, classDeclarationTree) -> {
-      Set<FunctionDeclarationTree> classMethods = new HashSet<>();
-      TreeVisitor<TreeContext> functionVisitor = new TreeVisitor<>();
-      functionVisitor.register(FunctionDeclarationTree.class,
-        (functionCtx, functionDeclarationTree) -> {
-          boolean isCurrentClassMethod = functionCtx.ancestors().stream()
-            .filter(ClassDeclarationTree.class::isInstance)
-            .findFirst().map(classDeclarationTree::equals)
-            .orElse(false);
-          if (isCurrentClassMethod) {
-            classMethods.add(functionDeclarationTree);
-          }
-        });
-      functionVisitor.scan(new TreeContext(), classDeclarationTree);
+      // return if this is not the outermost class
+      if (ctx.ancestors().stream().anyMatch(ClassDeclarationTree.class::isInstance)) {
+        return;
+      }
 
-      Set<IdentifierTree> usedIdentifiers = getAllUsedIdentifiers(ctx, classDeclarationTree);
+      Set<FunctionDeclarationTree> methods = new HashSet<>();
+      Set<IdentifierTree> usedIdentifiers = new HashSet<>();
 
-      usedIdentifiers.removeAll(classMethods.stream()
+      classDeclarationTree.descendants().forEach(tree -> {
+        if (tree instanceof FunctionDeclarationTree) {
+          methods.add(((FunctionDeclarationTree) tree));
+        } else if (tree instanceof IdentifierTree) {
+          usedIdentifiers.add(((IdentifierTree) tree));
+        }
+      });
+
+      usedIdentifiers.removeAll(methods.stream()
         .map(FunctionDeclarationTree::name)
         .collect(Collectors.toSet()));
 
-      classMethods.stream()
+      methods.stream()
```

Comment:
Here we loop methods.size() x usedIdentifierNames.size() times.
I would prefer to collect only IdentifierTree.identifier() string in a set, and use "contains" method.